File: typescript/tests/core/transparency-hook.test.ts
/**
 * Transparency-hook tests for `issueToken`.
 *
 * Covers (mirrors licensing/transparency_hook_test.go):
 *
 *   1. Hook fires once per successful issue with the right metadata.
 *   2. tokenSha256 matches an independent SHA-256 of the wire-token
 *      bytes — third-party logs can do the same hash and compare.
 *   3. Hook does NOT fire when issuance fails (wrong passphrase).
 *   4. Undefined hook is a zero-cost no-op.
 *   5. A hook that throws DOES propagate the throw — operators are
 *      expected to wrap their hook with try/catch if needed. This
 *      contract is pinned explicitly so a future "we should sandbox
 *      the hook" change is a deliberate decision, not an accidental
 *      drift.
 *   6. Concurrent issues fire the hook concurrently; each event is
 *      delivered exactly once with a distinct jti.
 */

import { describe, expect, it } from 'bun:test';
import { createHash } from 'node:crypto';

import { MemoryStorage } from '@anorebel/licensing/storage/memory';
import type { SignatureBackend } from '../../src/crypto/types.ts';
import type { Clock } from '../../src/id.ts';
import {
  createAdvancingClock,
  createLicense,
  ed25519Backend,
  generateRootKey,
  issueInitialSigningKey,
  issueToken,
  registerUsage,
  type TokenIssuedEvent,
} from '../../src/index.ts';
import type { Storage } from '../../src/storage/index.ts';
import type { KeyAlg, License, LicenseUsage } from '../../src/types.ts';

function mkBackends(): ReadonlyMap<KeyAlg, SignatureBackend> {
  return new Map<KeyAlg, SignatureBackend>([['ed25519', ed25519Backend]]);
}

async function setup(): Promise<{
  s: Storage;
  clock: Clock;
  license: License;
  usage: LicenseUsage;
  signingKid: string;
}> {
  const clock = createAdvancingClock('2026-05-01T10:00:00.000000Z');
  const s = new MemoryStorage({ clock });
  const root = await generateRootKey(s, clock, mkBackends(), {
    scope_id: null,
    alg: 'ed25519',
    passphrase: 'root-pw',
  });
  const signing = await issueInitialSigningKey(s, clock, mkBackends(), {
    scope_id: null,
    alg: 'ed25519',
    rootKid: root.kid,
    rootPassphrase: 'root-pw',
    signingPassphrase: 'sign-pw',
  });
  const initial = await createLicense(s, clock, {
    scope_id: null,
    template_id: null,
    licensable_type: 'User',
    licensable_id: 'u-1',
    max_usages: 5,
  });
  // registerUsage activates the license on first use; we want the
  // post-registration row (status=active) for issueToken.
  const reg = await registerUsage(s, clock, {
    license_id: initial.id,
    fingerprint: 'a'.repeat(64),
  });
  return { s, clock, license: reg.license, usage: reg.usage, signingKid: signing.kid };
}

describe('issueToken — transparency hook', () => {
  it('fires once with the right metadata on successful issue', async () => {
    const { s, clock, license, usage, signingKid } = await setup();
    const captured: TokenIssuedEvent[] = [];
    const result = await issueToken(s, clock, mkBackends(), {
      license,
      usage,
      ttlSeconds: 3600,
      alg: 'ed25519',
      signingPassphrase: 'sign-pw',
      transparencyHook: (e) => captured.push(e),
    });
    expect(captured).toHaveLength(1);
    const ev = captured[0] as TokenIssuedEvent;
    expect(ev.jti).toBe(result.jti);
    expect(ev.licenseId).toBe(license.id);
    expect(ev.usageId).toBe(usage.id);
    expect(ev.kid).toBe(signingKid);
    expect(ev.iat).toBe(result.iat);
    expect(ev.exp).toBe(result.exp);
  });

  it('tokenSha256 matches an independent SHA-256 of the wire token', async () => {
    const { s, clock, license, usage } = await setup();
    let captured: TokenIssuedEvent | null = null;
    const result = await issueToken(s, clock, mkBackends(), {
      license,
      usage,
      ttlSeconds: 3600,
      alg: 'ed25519',
      signingPassphrase: 'sign-pw',
      transparencyHook: (e) => {
        captured = e;
      },
    });
    const want = createHash('sha256').update(result.token).digest('hex').toLowerCase();
    expect(captured).not.toBeNull();
    expect((captured as unknown as TokenIssuedEvent).tokenSha256).toBe(want);
    // Format invariant — 64 lowercase hex chars.
    expect((captured as unknown as TokenIssuedEvent).tokenSha256).toMatch(/^[0-9a-f]{64}$/);
  });

  it('does not fire when issue fails (wrong passphrase)', async () => {
    const { s, clock, license, usage } = await setup();
    let called = false;
    await expect(
      issueToken(s, clock, mkBackends(), {
        license,
        usage,
        ttlSeconds: 3600,
        alg: 'ed25519',
        signingPassphrase: 'wrong-passphrase',
        transparencyHook: () => {
          called = true;
        },
      }),
    ).rejects.toBeDefined();
    expect(called).toBe(false);
  });

  it('undefined hook is a zero-cost no-op', async () => {
    const { s, clock, license, usage } = await setup();
    const result = await issueToken(s, clock, mkBackends(), {
      license,
      usage,
      ttlSeconds: 3600,
      alg: 'ed25519',
      signingPassphrase: 'sign-pw',
      // transparencyHook: undefined
    });
    expect(result.token.startsWith('LIC1.')).toBe(true);
  });

  it('throwing hook propagates — sandboxing is the operator wrapper job', async () => {
    const { s, clock, license, usage } = await setup();
    await expect(
      issueToken(s, clock, mkBackends(), {
        license,
        usage,
        ttlSeconds: 3600,
        alg: 'ed25519',
        signingPassphrase: 'sign-pw',
        transparencyHook: () => {
          throw new Error('hook explosion');
        },
      }),
    ).rejects.toThrow(/hook explosion/);
  });

  it('concurrent issues deliver distinct events with distinct jtis', async () => {
    // Setup with a license that has enough seat headroom for N parallel usages.
    const clock = createAdvancingClock('2026-05-01T10:00:00.000000Z');
    const s = new MemoryStorage({ clock });
    const root = await generateRootKey(s, clock, mkBackends(), {
      scope_id: null,
      alg: 'ed25519',
      passphrase: 'root-pw',
    });
    await issueInitialSigningKey(s, clock, mkBackends(), {
      scope_id: null,
      alg: 'ed25519',
      rootKid: root.kid,
      rootPassphrase: 'root-pw',
      signingPassphrase: 'sign-pw',
    });
    const N = 20;
    const initialLicense = await createLicense(s, clock, {
      scope_id: null,
      template_id: null,
      licensable_type: 'User',
      licensable_id: 'u-1',
      max_usages: N,
    });
    // Fingerprint is exactly 64 hex chars — pad the index into the
    // last few positions so each is distinct.
    const usages: LicenseUsage[] = [];
    let activated: License | null = null;
    for (let i = 0; i < N; i++) {
      const fp = ('a'.repeat(60) + i.toString(16).padStart(4, '0')).slice(0, 64);
      const reg = await registerUsage(s, clock, { license_id: initialLicense.id, fingerprint: fp });
      usages.push(reg.usage);
      activated = reg.license;
    }
    const license = activated as License;

    const seenJtis = new Set<string>();
    let dupSeen = false;
    const hook = (e: TokenIssuedEvent): void => {
      if (seenJtis.has(e.jti)) dupSeen = true;
      seenJtis.add(e.jti);
    };

    await Promise.all(
      usages.map((u) =>
        issueToken(s, clock, mkBackends(), {
          license,
          usage: u,
          ttlSeconds: 3600,
          alg: 'ed25519',
          signingPassphrase: 'sign-pw',
          transparencyHook: hook,
        }),
      ),
    );
    expect(dupSeen).toBe(false);
    expect(seenJtis.size).toBe(N);
  });
});

File: licensing/transparency_hook_test.go
package licensing_test

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"regexp"
	"sync"
	"testing"

	lic "github.com/AnoRebel/licensing/licensing"
	ed "github.com/AnoRebel/licensing/licensing/crypto/ed25519"
	"github.com/AnoRebel/licensing/licensing/storage/memory"
)

// Transparency-hook tests. Cover:
//
//   1. Hook fires once per successful issue with the right metadata.
//   2. Hook receives a SHA-256 hash of the wire token bytes (64-char
//      lowercase hex). Computing the hash independently and comparing
//      proves the contract — a third-party log can do the same.
//   3. Hook does NOT fire on issuance failure (e.g. wrong passphrase).
//   4. Nil hook is a zero-cost no-op.
//   5. A panicking hook does NOT bring down the issuer.
//   6. Concurrent issues fire the hook concurrently; each event is
//      delivered exactly once.

func freshFixture(t *testing.T) (lic.Storage, *lic.AlgorithmRegistry, *lic.License, *lic.LicenseUsage, string) {
	t.Helper()
	const passphrase = "pp-test-pp-test-pp-test-pp-test-"
	storage := memory.New(memory.Options{})
	reg := lic.NewAlgorithmRegistry()
	if err := reg.Register(ed.New()); err != nil {
		t.Fatal(err)
	}
	clk := lic.SystemClock{}

	root, err := lic.GenerateRootKey(storage, clk, reg,
		lic.GenerateRootKeyInput{
			Alg:        lic.AlgEd25519,
			Passphrase: passphrase,
		},
		lic.KeyIssueOptions{Actor: "test"},
	)
	if err != nil {
		t.Fatalf("generate root: %v", err)
	}
	if _, err := lic.IssueInitialSigningKey(storage, clk, reg,
		lic.IssueInitialSigningKeyInput{
			RootKid:           root.Kid,
			Alg:               lic.AlgEd25519,
			RootPassphrase:    passphrase,
			SigningPassphrase: passphrase,
		},
		lic.KeyIssueOptions{Actor: "test"},
	); err != nil {
		t.Fatalf("issue signing: %v", err)
	}

	license, err := storage.CreateLicense(lic.LicenseInput{
		LicensableType: "User",
		LicensableID:   "u-1",
		LicenseKey:     "LIC-AAAA-BBBB-CCCC-DDDD",
		Status:         lic.LicenseStatusActive,
		MaxUsages:      5,
	})
	if err != nil {
		t.Fatalf("create license: %v", err)
	}
	usage, err := storage.CreateUsage(lic.LicenseUsageInput{
		LicenseID:    license.ID,
		Fingerprint:  "fp-test",
		Status:       lic.UsageStatusActive,
		RegisteredAt: clk.NowISO(),
	})
	if err != nil {
		t.Fatalf("create usage: %v", err)
	}
	return storage, reg, license, usage, passphrase
}

func TestTransparencyHook_FiresOnceWithCorrectMetadata(t *testing.T) {
	storage, reg, license, usage, passphrase := freshFixture(t)
	clk := lic.SystemClock{}

	var captured []lic.TokenIssuedEvent
	var mu sync.Mutex
	hook := func(ev lic.TokenIssuedEvent) {
		mu.Lock()
		defer mu.Unlock()
		captured = append(captured, ev)
	}

	res, err := lic.IssueToken(storage, clk, reg, lic.IssueTokenInput{
		License:           license,
		Usage:             usage,
		TTLSeconds:        3600,
		Alg:               lic.AlgEd25519,
		SigningPassphrase: passphrase,
		TransparencyHook:  hook,
	})
	if err != nil {
		t.Fatalf("issue: %v", err)
	}

	if len(captured) != 1 {
		t.Fatalf("hook should fire exactly once; got %d", len(captured))
	}
	ev := captured[0]
	if ev.Jti != res.Jti {
		t.Errorf("Jti drift: hook=%s result=%s", ev.Jti, res.Jti)
	}
	if ev.LicenseID != license.ID {
		t.Errorf("LicenseID drift: %s vs %s", ev.LicenseID, license.ID)
	}
	if ev.UsageID != usage.ID {
		t.Errorf("UsageID drift: %s vs %s", ev.UsageID, usage.ID)
	}
	if ev.Kid != res.Kid {
		t.Errorf("Kid drift: %s vs %s", ev.Kid, res.Kid)
	}
	if ev.Iat != res.Iat || ev.Exp != res.Exp {
		t.Errorf("iat/exp drift: hook=(%d,%d) result=(%d,%d)",
			ev.Iat, ev.Exp, res.Iat, res.Exp)
	}
}

func TestTransparencyHook_TokenSHA256MatchesIndependentHash(t *testing.T) {
	// The contract: TokenSHA256 is sha256(wire_token_bytes), lowercase
	// hex, 64 chars. A third-party log doing its own hash of the
	// token bytes MUST produce the same string.
	storage, reg, license, usage, passphrase := freshFixture(t)
	clk := lic.SystemClock{}

	var captured lic.TokenIssuedEvent
	hook := func(ev lic.TokenIssuedEvent) { captured = ev }

	res, err := lic.IssueToken(storage, clk, reg, lic.IssueTokenInput{
		License:           license,
		Usage:             usage,
		TTLSeconds:        3600,
		Alg:               lic.AlgEd25519,
		SigningPassphrase: passphrase,
		TransparencyHook:  hook,
	})
	if err != nil {
		t.Fatalf("issue: %v", err)
	}

	// Independent hash of the wire token.
	want := sha256.Sum256([]byte(res.Token))
	wantHex := hex.EncodeToString(want[:])

	if captured.TokenSHA256 != wantHex {
		t.Fatalf("TokenSHA256 drift: hook=%s independent=%s",
			captured.TokenSHA256, wantHex)
	}

	// Format invariant — 64 lowercase hex chars.
	if !regexp.MustCompile(`^[0-9a-f]{64}$`).MatchString(captured.TokenSHA256) {
		t.Fatalf("TokenSHA256 not lowercase hex/64 chars: %q", captured.TokenSHA256)
	}
}

func TestTransparencyHook_DoesNotFireOnIssueFailure(t *testing.T) {
	storage, reg, license, usage, _ := freshFixture(t)
	clk := lic.SystemClock{}

	called := false
	hook := func(lic.TokenIssuedEvent) { called = true }

	_, err := lic.IssueToken(storage, clk, reg, lic.IssueTokenInput{
		License:           license,
		Usage:             usage,
		TTLSeconds:        3600,
		Alg:               lic.AlgEd25519,
		SigningPassphrase: "wrong-passphrase-wrong-passphrase-",
		TransparencyHook:  hook,
	})
	if err == nil {
		t.Fatal("expected wrong-passphrase to fail")
	}
	if called {
		t.Fatal("hook must NOT fire when issue fails")
	}
}

func TestTransparencyHook_NilHookIsZeroCost(t *testing.T) {
	// Sanity test: nil hook leaves the API and behaviour unchanged.
	storage, reg, license, usage, passphrase := freshFixture(t)
	clk := lic.SystemClock{}

	res, err := lic.IssueToken(storage, clk, reg, lic.IssueTokenInput{
		License:           license,
		Usage:             usage,
		TTLSeconds:        3600,
		Alg:               lic.AlgEd25519,
		SigningPassphrase: passphrase,
		// TransparencyHook: nil
	})
	if err != nil {
		t.Fatalf("issue with nil hook: %v", err)
	}
	if res.Token == "" {
		t.Fatal("token should still be returned with nil hook")
	}
}

func TestTransparencyHook_PanicDoesNotEscapeIssuer(t *testing.T) {
	// A misbehaving hook MUST NOT take down the issuer. The current
	// behaviour propagates the panic — operators are expected to wrap
	// their hook with recover() if needed. This test pins the
	// contract explicitly so a future "we should sandbox the hook"
	// change is a deliberate decision, not an accidental drift.
	storage, reg, license, usage, passphrase := freshFixture(t)
	clk := lic.SystemClock{}

	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected hook panic to propagate (operator responsibility)")
		}
	}()

	_, _ = lic.IssueToken(storage, clk, reg, lic.IssueTokenInput{
		License:           license,
		Usage:             usage,
		TTLSeconds:        3600,
		Alg:               lic.AlgEd25519,
		SigningPassphrase: passphrase,
		TransparencyHook: func(lic.TokenIssuedEvent) {
			panic(errors.New("hook panic"))
		},
	})
	t.Fatal("unreachable — defer recover should have caught the panic above")
}

func TestTransparencyHook_ConcurrentIssuesDeliverDistinctEvents(t *testing.T) {
	storage, reg, license, _, passphrase := freshFixture(t)
	clk := lic.SystemClock{}

	// Each goroutine creates its own usage so the issuance succeeds.
	const N = 25
	usages := make([]*lic.LicenseUsage, N)
	for i := 0; i < N; i++ {
		u, err := storage.CreateUsage(lic.LicenseUsageInput{
			LicenseID:   license.ID,
			Fingerprint: regexp.MustCompile(`\W`).ReplaceAllString("fp-"+string(rune('a'+i)), ""),
			Status:      lic.UsageStatusActive,
		})
		if err != nil {
			t.Fatal(err)
		}
		usages[i] = u
	}

	var mu sync.Mutex
	seen := make(map[string]bool)
	hook := func(ev lic.TokenIssuedEvent) {
		mu.Lock()
		defer mu.Unlock()
		if seen[ev.Jti] {
			t.Errorf("duplicate jti delivered: %s", ev.Jti)
		}
		seen[ev.Jti] = true
	}

	var wg sync.WaitGroup
	for i := 0; i < N; i++ {
		wg.Add(1)
		go func(u *lic.LicenseUsage) {
			defer wg.Done()
			_, err := lic.IssueToken(storage, clk, reg, lic.IssueTokenInput{
				License:           license,
				Usage:             u,
				TTLSeconds:        3600,
				Alg:               lic.AlgEd25519,
				SigningPassphrase: passphrase,
				TransparencyHook:  hook,
			})
			if err != nil {
				t.Errorf("issue: %v", err)
			}
		}(usages[i])
	}
	wg.Wait()

	if len(seen) != N {
		t.Fatalf("expected %d distinct jtis, got %d", N, len(seen))
	}
}
